Use advisory lock during table creation with knex

lexuzieel opened the pull request:
Hello, I have encountered a somewhat weird edge case which this PR aims to fix.

I am currently developing a package that uses Verrou under the hood. I am using `knex` adapter for postgres and in my project I am using [testcontainers](https://node.testcontainers.org/modules/postgresql/) to spin up ephemeral postgres instance.

Somehow during testing there occurs a race condition where Verrou tries to create table twice & as a result it tries to create primary key multiple times as well. Here is the error that I get:

```
error: create table "verrou" ("key" varchar(255) not null,
"owner" varchar(255) not null, "expiration" bigint null,
constraint "verrou_pkey" primary key ("key")) - duplicate
key value violates unique constraint "pg_type_typname_nsp_index"
```

Wrapping table creation logic into a transaction and issuing an advisory lock fixes this for my case.

Take note, that I have not tested it with `kysely` but I think it is the same with that one.

The only caveat with this PR is the hash generation for cache (I am not that familiar with postgres) but I suppose it should be safe since it is inside transaction.

lexuzieel committed:
fix(knex): use advisory lock during table creation

Julien-R44 commented:
Hey, thanks a lot for investigating

what about using `CREATE TABLE IF NOT EXIST` instead? i think it should also solve your issue? Maybe you already tried

lexuzieel commented:
Yeah I tried using it briefly through `.raw()`, however it didn't have any effect. I think this might be because primary index creation is a separate operation.

I tried using raw SQL like so:

```
error: 
      CREATE TABLE IF NOT EXISTS verrou (
        key varchar(255) NOT NULL PRIMARY KEY,
        owner varchar(255) NOT NULL,
        expiration bigint NULL
      )
     - duplicate key value violates unique constraint "pg_type_typname_nsp_index"
```

Julien-R44 commented:
Hey man, I attempted a fix in the commit jsut above. your solution could indeed work, but issue is that its Postgres-specific, and the Knex driver is used with other databases too

So lets see if the next release fixes your issue. If not, a reproduction would be super helpful!

lexuzieel commented:
Awesome, looks like your fix works!

I've updated the package to 0.5.2 and added some debug logging to verify - catch branch executes, but the code does not throw anymore, so it works as intended.

Anyway, I have created a reproduction branch in case you are curious: https://github.com/lexuzieel/lavoro/tree/verrou-knex-race

Since the test is flaky, you better run it like so (it will automatically spin up postgres in docker):

```
for i in {1..100}; do npm run test postgres_lock_provider; done
```

This is an error that I was getting:

<img width="1246" height="663" alt="изображение" src="https://github.com/user-attachments/assets/8c0e2177-2e81-4fb9-a87e-88e5c0ea17b6" />

---

P.S. Yes, the name of the package is in fact inspired by verrou in a way :) and the package has adonisjs integration. Actually, I have in my plans to open a discussion with you guys in adonisjs forum, but after I draft up some docs & motivation.
